### PR TITLE
Fix LICENSE/LICENSE.bin for release

### PR DESCRIPTION
**Pull request by jbonofre**
- Fix `LICENSE` to be included in source distribution
- Add `LICENSE-BINARY-DIST` and include in the binary distributions

### CODE REVIEW

File: polaris-service/build.gradle.kts [hunk 1]
```diff
@@ -41,6 +41,11 @@ dependencies {
     exclude("ch.qos.reload4j", "reload4j")
     exclude("log4j", "log4j")
     exclude("org.apache.zookeeper", "zookeeper")
+    exclude("org.apache.hadoop.thirdparty", "hadoop-shaded-protobuf_3_25")
```
Comment (snazy) [inline polaris-service/build.gradle.kts:44]: Hadoop's an Apache project, why these excludes?

File: LICENSE [hunk 2]
```diff
@@ -200,6713 +200,9 @@
...
-
-NO WARRANTY
-
-11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.
-
-12. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED
```
Comment (snazy) [inline LICENSE:208]: Isn't this included in the sentence above? `gradlew` is AL2 - and `gradle-wrapper.properties` is actually "ours".

Comment (snazy) [reply LICENSE:208]: GH placed the comment wrong - it's for the "Gradle sentence" at the top

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -18,9 +18,9 @@
 #
 
 [versions]
-hadoop = "3.4.0"
+hadoop = "3.4.1"
```
Comment (snazy) [inline gradle/libs.versions.toml:21]: Can we do the version bumps in a separate PR?

Should actually be done by renovate ;)

File: build-logic/src/main/kotlin/LicenseFileValidation.kt [hunk 4]
```diff
@@ -58,7 +58,7 @@ class LicenseFileValidation : DependencyFilter {
   override fun filter(data: ProjectData?): ProjectData {
     data!!
 
-    val rootLicenseFile = data.project.rootProject.file("LICENSE").readText()
+    val rootLicenseFile = data.project.rootProject.file("LICENSE.bin").readText()
```
Comment (snazy) [inline build-logic/src/main/kotlin/LicenseFileValidation.kt:61]: `.bin` implies it's a binary file - maybe `DEPENDENCY-LICENSES.txt` or just `DEPENDENCY-LICENSES`?

File: polaris-service/build.gradle.kts (hunk 1 above)
Comment (jbonofre) [reply polaris-service/build.gradle.kts:44]: I excluded because it's not actually used in Polaris.

File: gradle/libs.versions.toml (hunk 3 above)
Comment (jbonofre) [reply gradle/libs.versions.toml:21]: Sure, I can do in a separate PR. The reason why I included here it's because the transitive dependencies changed a bit.

File: build-logic/src/main/kotlin/LicenseFileValidation.kt (hunk 4 above)
Comment (jbonofre) [reply build-logic/src/main/kotlin/LicenseFileValidation.kt:61]: That's correct: `LICENSE.bin` is targeted to be included in binary distributions (like `polaris-service` tar/zip).
No problem to rename as `LICENSE-BINARY-DEPENDENCY` for instance.

File: LICENSE.bin [hunk 5]
```diff
@@ -0,0 +1,4712 @@
...
+  included in all copies or substantial portions of the Software.
+
+  THE  SOFTWARE IS  PROVIDED  "AS  IS", WITHOUT  WARRANTY  OF ANY  KIND,
+  EXPRESS OR  IMPLIED, INCLUDING  BUT NOT LIMITED  TO THE  WARRANTIES OF
+  MERCHANTABILITY,    FITNESS    FOR    A   PARTICULAR    PURPOSE    AND
+  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+  OF CONTRACT, TORT OR OTHERWISE,  ARISING FROM, OUT OF OR IN CONNECTION
+  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
```
Comment (flyrain) [inline LICENSE.bin:4678]: Do we distributed the third party jar file(e.g., `org.slf4j:slf4j-api`)?  If yes, then we don't have to copy the license in Polaris. The jar has the license. 
Screenshot 2024-12-03 at 9 53 54 AM

Comment (flyrain) [reply LICENSE.bin:4678]: if we have to include the third party license, can we use the link to save the space like what Iceberg does?
```
This product includes code from Apache Hive.

* Hive metastore derby schema in hive-schema-3.1.0.derby.sql

Copyright: 2011-2018 The Apache Software Foundation
Home page: https://hive.apache.org/
License: https://www.apache.org/licenses/LICENSE-2.0
```
https://github.com/apache/iceberg/blob/feed4e2544b5839fbc2fe040965af3906d053302/LICENSE#L273-L273

File: build-logic/src/main/kotlin/LicenseFileValidation.kt (hunk 4 above)
Comment (snazy) [reply build-logic/src/main/kotlin/LicenseFileValidation.kt:61]: Then let's rename `LICENSE.bin` to `DEPENDENCY-LICENSE`, because `.bin` usually means "this" is a binary file.

Comment (flyrain) [reply build-logic/src/main/kotlin/LicenseFileValidation.kt:61]: This PR creates a new license file. Do we need two license files. I think the one including the dependencies license is good enough.

Comment (jbonofre) [reply build-logic/src/main/kotlin/LicenseFileValidation.kt:61]: Yes we need two because we have a source distribution and a binary distribution.

If we don't distribute the binary distribution then only LICENSE would be enough. But as we have the polaris-service tar, to distribute it, we need the license for binaries included in this distribution.

Comment (flyrain) [reply build-logic/src/main/kotlin/LicenseFileValidation.kt:61]: For binary distribution, do we need to aggregate all third-party licenses into a text file, as the third party jar files have them already?

File: LICENSE.bin (hunk 5 above)
Comment (jbonofre) [reply LICENSE.bin:4678]: We ship slf4j in the binary distribution. No need to include the License only copy the notice content (according to alv2 4d section). 

Iceberg is different as it's only source distribution (not binary). So it mentions only the source code coming from other projects. 

Just to be clear: I'm pretty sure of my changes here (especially for the source distribution). To license and notice of the binary distribution are generated by the gradle plugin.

Comment (jbonofre) [reply LICENSE.bin:4678]: > Do we distributed the third party jar file(e.g., `org.slf4j:slf4j-api`)?  If yes, then we don't have to copy the license in Polaris. The jar has the license. 
> 
> Screenshot 2024-12-03 at 9 53 54 AM
> 
> 

That doesn't matter: the notice would have been important but slf4j doesn't have notice

Comment (flyrain) [reply LICENSE.bin:4678]: > Iceberg is different as it's only source distribution (not binary). So it mentions only the source code coming from other projects.

Should we do that for the Polaris source distribution? Oh, I guess this statement does the job. We should be fine?
```
   This product bundles various third-party components also under the
   Apache Software License 2.0.
```

**Review by flyrain — APPROVED** [review]
LGTM. Thanks @jbonofre for driving this.

**Comment by jbonofre** [comment]
@snazy @flyrain I updated the PR with your comments.

**Review by flyrain — APPROVED** [review]
+1

File: polaris-service/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,290 @@
+/*
```
[the remark this answers is not in the document]
Comment (jbonofre) [reply polaris-service/build.gradle.kts:1]: Good catch, it's probably due to the rebase. I'm cleaning up !

**Comment by jbonofre** [comment]
@snazy I nuked the `polaris-service/build.gradle.kts` and rebase

File: dropwizard/service/build.gradle.kts [hunk 7]
```diff
@@ -183,4 +183,13 @@ tasks.register<Sync>("prepareDockerDist") {
 
 tasks.named("build").configure { dependsOn("prepareDockerDist") }
 
+distributions {
```
Comment (snazy) [inline dropwizard/service/build.gradle.kts:186]: Hm - it looks like none of the dist(s) actually works.

```
./gradlew :polaris-dropwizard-service:assembleDist
```
generates TWO sets of distributions:
```
$ ls -1 dropwizard/service/build/distributions/
polaris-dropwizard-service-1.0.0-incubating-SNAPSHOT-license-report.zip
polaris-dropwizard-service-1.0.0-incubating-SNAPSHOT.tar
polaris-dropwizard-service-1.0.0-incubating-SNAPSHOT.zip
polaris-dropwizard-service-shadow-1.0.0-incubating-SNAPSHOT.tar
polaris-dropwizard-service-shadow-1.0.0-incubating-SNAPSHOT.zip
```

The "non shadow" one has the NOTICE & LICENSE files, but running `polaris-dropwizard-service-1.0.0-incubating-SNAPSHOT/bin/polaris-service` leads to 
```
Error: Could not find or load main class org.apache.polaris.service.dropwizard.PolarisApplication
Caused by: java.lang.ClassNotFoundException: org.apache.polaris.service.dropwizard.PolarisApplication
```

The "shadow" does not have the NOTICE + LICENSE files, but running `polaris-dropwizard-service-shadow-1.0.0-incubating-SNAPSHOT/bin/polaris-dropwizard-service` does something.

Comment (jbonofre) [reply dropwizard/service/build.gradle.kts:186]: Yes, the `polaris-service` script never worked (see #542 ). It's not related to this PR, it's like this for a while.

Good point about "shadow" distribution. I actually wonder if you need both (only one binary distribution is enough).

Comment (snazy) [reply dropwizard/service/build.gradle.kts:186]: Frankly speaking, I think it's even better to get #469 in and let it replace the DW stuff entirely. Especially given the amount of issues.

**Review by snazy — APPROVED** [review]

**Comment by flyrain** [comment]
Thanks @jbonofre for the PR, thanks @snazy for the reivew.
